### PR TITLE
Update hlint

### PR DESCRIPTION
**Pull request by shayne-fletcher**
Update hlint. Depends on https://github.com/digital-asset/hlint/pull/52

### CODE REVIEW

File: compiler/damlc/tests/daml-test-files/Newtype.daml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 -- Copyright (c) 2020, Digital Asset (Switzerland) GmbH and/or its affiliates.
 -- All rights reserved.
 
+-- @INFO Eta reduce
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/Newtype.daml]: I’m not really a fan of the `eta reduce` hint. I think it is more confusing than helpful. Can we disable it by default? cc @hurryabit

File: compiler/damlc/tests/daml-test-files/Generics.daml [hunk 2]
```diff
@@ -1,6 +1,8 @@
 -- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
 -- SPDX-License-Identifier: Apache-2.0
 
+-- @INFO Unused LANGUAGE pragma
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/Generics.daml]: Which one is unused? Can we just remove the unnecessary pragma here please?

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 3]
```diff
@@ -117,12 +117,12 @@
...
     - hint: {lhs: length x /= 0, rhs: not (null x), name: Use null}
     - hint: {lhs: 0 /= length x, rhs: not (null x), name: Use null}
-    - hint: {lhs: "\\x -> [x]", rhs: "(:[])", name: "Use :"}
+    - hint: {lhs: "\\x -> [x]", rhs: "(::[])", name: "Use :"}
     - warn: {lhs: map (uncurry f) (zip x y), rhs: zipWith f x y}
     - hint: {lhs: map f (zip x y), rhs: zipWith (curry f) x y, side: isVar f}
     - warn: {lhs: not (elem x y), rhs: notElem x y}
     - hint: {lhs: foldr f z (map g x), rhs: foldr (f . g) z x, name: Fuse foldr/map}
-    - warn: {lhs: "x ++ concatMap (' ':) y", rhs: "unwords (x:y)"}
+    - warn: {lhs: "x ++ concatMap (' '::) y", rhs: "unwords (x::y)"}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]: That hint is wrong. `unwords` operates on `Text` which is not a list in `DAML`.

File: compiler/damlc/daml-ide-core/dlint.yaml [hunk 4]
```diff
@@ -117,12 +117,12 @@
     - hint: {lhs: "[] == x", rhs: null x}
     - hint: {lhs: length x /= 0, rhs: not (null x), name: Use null}
     - hint: {lhs: 0 /= length x, rhs: not (null x), name: Use null}
-    - hint: {lhs: "\\x -> [x]", rhs: "(:[])", name: "Use :"}
+    - hint: {lhs: "\\x -> [x]", rhs: "(::[])", name: "Use :"}
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/dlint.yaml]:
```suggestion
    - hint: {lhs: "\\x -> [x]", rhs: "(::[])", name: "Use ::"}
```

File: compiler/damlc/tests/daml-test-files/TextMap.daml [hunk 5]
```diff
@@ -6,7 +6,7 @@
 -- @INFO Use /=
 -- @INFO Use /=
 -- @INFO Use /=
-
+-- @INFO Avoid lambda
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/TextMap.daml]: Similar comment as for the eta reduce rule. I don’t think encouraging point-free style is helpful for beginner DAML programmers.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 6]
```diff
@@ -99,7 +99,7 @@ import           Data.Data hiding (TyCon)
 import qualified Data.Decimal as Decimal
 import           Data.Foldable (foldlM)
 import           Data.Int
-import           Data.List.Extra hiding (for)
```
[the remark this answers is not in the document]
Comment (shayne-fletcher) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: It no longer exists a function called `for`.

File: compiler/damlc/tests/daml-test-files/Newtype.daml (hunk 1 above)
Comment (shayne-fletcher) [reply compiler/damlc/tests/daml-test-files/Newtype.daml]: Done.

File: compiler/damlc/tests/daml-test-files/Generics.daml (hunk 2 above)
Comment (shayne-fletcher) [reply compiler/damlc/tests/daml-test-files/Generics.daml]: Done

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 3 above)
Comment (shayne-fletcher) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Removed the rule

File: compiler/damlc/daml-ide-core/dlint.yaml (hunk 4 above)
Comment (shayne-fletcher) [reply compiler/damlc/daml-ide-core/dlint.yaml]: Done

File: compiler/damlc/tests/daml-test-files/TextMap.daml (hunk 5 above)
Comment (shayne-fletcher) [reply compiler/damlc/tests/daml-test-files/TextMap.daml]: Globally disabled

**Comment by shayne-fletcher** [comment]
@cocreature PR adjusted as per your comments.

File: compiler/damlc/tests/daml-test-files/Newtype.daml (hunk 1 above)
Comment (hurryabit) [reply compiler/damlc/tests/daml-test-files/Newtype.daml]: I agree that's not a particularly helpful rule.

File: compiler/damlc/tests/daml-test-files/TextMap.daml (hunk 5 above)
Comment (hurryabit) [reply compiler/damlc/tests/daml-test-files/TextMap.daml]: I think point-free style is almost never good. 😸

File: compiler/damlc/tests/daml-test-files/ActionFail.daml [hunk 7]
```diff
@@ -1,6 +1,8 @@
 -- Copyright (c) 2020, Digital Asset (Switzerland) GmbH and/or its affiliates.
 -- All rights reserved.
 
+-- @INFO Use catOptionals
+
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/ActionFail.daml]: What are we testing here? And why is this test called `ActionFail`?

Comment (shayne-fletcher) [reply compiler/damlc/tests/daml-test-files/ActionFail.daml]: That's lost in the mists of time but if I had to guess, it's that pattern match failures work correctly in the presence of `RebindableSyntax` (https://github.com/ghc-proposals/ghc-proposals/pull/168).

**Pull request merged**